executor/src/witgen/generator.rs

```diff
@@ -418,10 +464,14 @@ where
 
     /// Processes the evaluation result: Stores failure reasons and updates next values.
     /// Returns true if a new value or constraint was determined.
-    fn handle_eval_result(&mut self, result: EvalResult<T>) -> bool {
+    fn handle_eval_result(&mut self, result: EvalResult<T>, strategy: SolvingStrategy) -> bool {
         match result {
             Ok(constraints) => {
                 let progress = !constraints.is_empty();
+                // If we assume unknown variables to be zero, we cannot learn anything new.
+                if strategy == SolvingStrategy::AssumeZero {
+                    assert!(!progress);
+                }
```

Comment:
Much less readable one-liner version that I would have used:
```
assert!(strategy != SolvingStrategy::AssumeZero || !progress);
```
or maybe even
```
assert!(!(strategy == SolvingStrategy::AssumeZero && progress));
```
Glad you didn't!